subsys/usb/class/audio/audio.c

```diff
@@ -10,3 +10,952 @@
+ * If so then true is returned and audio_dev_data is considered correct device
+ * data.
+ *
+ * @param [in] audio_dev_data USB audio device data.
+ * @param [in] interface      USB Audio interface number.
+ *
+ * @return true if interface matched audio_dev_data, false otherwise.
+ */
+static bool is_interface_valid(struct usb_audio_dev_data_t *audio_dev_data,
+			       u8_t interface)
+{
+	const struct cs_ac_interface_descriptor_header *header;
+
+	header = audio_dev_data->header_descr;
+	u8_t desc_iface = 0;
+
+	for (size_t i = 0; i < header->bInCollection; i++) {
+		desc_iface = header->baInterfaceNr[i];
+		if (desc_iface == interface) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+/**
+ * @brief Helper funciton for getting the audio_dev_data by the interface
+ *	  number.
+ *
+ * This function searches through all audio devices the one with given
+ * interface number and returns the audio_dev_data structure for this device.
+ *
+ * @param [in] interface USB Audio interface addressed by the request.
+ *
+ * @return audio_dev_data for given interface, NULL if not found.
+ */
+static struct usb_audio_dev_data_t *get_audio_dev_data_by_interface(
+								u8_t interface)
+{
+	struct usb_dev_data *dev_data;
+	struct usb_audio_dev_data_t *audio_dev_data;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&usb_audio_data_devlist, dev_data, node) {
+		audio_dev_data = CONTAINER_OF(dev_data,
+					struct usb_audio_dev_data_t, common);
+
+		if (is_interface_valid(audio_dev_data, interface)) {
+			return audio_dev_data;
+		}
+	}
+	return NULL;
+}
+
+/**
+ * @brief Handler for feature unit requests.
+ *
+ * This function handles feature unit specific requests.
+ * If request is properly served 0 is returned. Negative errno
+ * is returned in case of an error. This leads to setting stall on IN EP0.
+ *
+ * @param [in]  dev_data USB audio device data.
+ * @param [in]  pSetup   Information about the executed request.
+ * @param [in]  len      Size of the buffer.
+ * @param [out] data     Buffer containing the request result.
+ *
+ * @return 0 if succesfulf, negative errno otherwise.
+ */
+static int handle_feature_unit_req(struct usb_audio_dev_data_t *audio_dev_data,
+				   struct usb_setup_packet *pSetup,
+				   s32_t *len,
+				   u8_t **data)
+{
+	enum usb_audio_feature_unit_control_selectors cs;
+	struct usb_audio_feature_unit_evt evt;
+	const struct feature_unit_descriptor *fu;
+	u8_t device;
+	u8_t fu_id;
+	u8_t ch;
+
+	fu_id = ((pSetup->wIndex) >> 8) & 0xFF;
+	fu = get_feature_unit(audio_dev_data, &device, fu_id);
+	ch = (pSetup->wValue) & 0xFF;
+	cs = ((pSetup->wValue) >> 8) & 0xFF;
+
+	u8_t dev_chs = audio_dev_data->ch_num[device];
+	u8_t *controls = audio_dev_data->controls[device];
+
+	LOG_DBG("CS: %d, CN: %d, len: %d", cs, ch, *len);
+
+	/* Error checking */
+	if (!(BIT(cs) & (get_controls(fu) << 1UL))) {
+		return -EINVAL;
+	} else if (ch >= dev_chs && ch != 0xFF) {
+		return -EINVAL;
+	}
+
+	switch (cs) {
+	case USB_AUDIO_FU_MUTE_CONTROL:
```

Comment:
```c
if (cs != USB_AUDIO_FU_MUTE_CONTROL) {
    return -ENOSUP;
}
```
You will probably expand it in the future but then you can move MUTE_CONTROL to a handle_fu_req_mute function.

Reply:
I will expand and yes I planned to create something like `handle_fu_req_mute()` in the future. I will leave `switch` as I think this will later be cleaner when I add `VOLUME_CONTROL` for example.

Reply:
This switch block will not cleaner then, please move it to `handle_fu_req_mute`. switch{switch{if{if}}} is ugly

Reply:
Are You suggesting this?
```
if (cs != USB_AUDIO_FU_MUTE_CONTROL) {
    return -ENOSUP;
}

return handle_fu_req_mute(...); 
```

Reply:
You can leave the first switch block, just move the mute control to handle_fu_req_mute(...);